google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/project/AppEngineProjectService.java

```diff
@@ -0,0 +1,123 @@
+import com.intellij.openapi.project.Project;
+import com.intellij.packaging.artifacts.Artifact;
+import com.intellij.psi.xml.XmlTag;
+import com.intellij.remoteServer.configuration.deployment.DeploymentSource;
+
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.Collection;
+
+/**
+ * A set of helper methods for inspecting an App Engine project's structure and configuration.
+ */
+public abstract class AppEngineProjectService {
+
+  public static final String APP_ENGINE_STANDARD_FACET_NAME = "Google App Engine";
+
+  public static AppEngineProjectService getInstance() {
+    return ServiceManager.getService(AppEngineProjectService.class);
+  }
+
+  /**
+   * Given a deployment source, returns the xml tag corresponding to the project's appengine-web.xml
+   * compat configuration or null if there isn't one.
+   */
+  @Nullable
+  public abstract XmlTag getFlexCompatXmlConfiguration(@NotNull Project project,
+      @NotNull DeploymentSource source);
+
+  /**
+   * @see AppEngineProjectService#getFlexCompatXmlConfiguration(Project, DeploymentSource).
+   */
+  @Nullable
+  public abstract XmlTag getFlexCompatXmlConfiguration(@NotNull Project project,
+      @Nullable Artifact artifact);
+
+  /**
+   * Determines if a project is set up like an App Engine standard project but is configured in
+   * 'compatibility' mode. This indicates that the project runs in the flexible environment.
+   *
+   * <p>A flex compat project has an appengine-web.xml with either:
+   * {@code
+   * <vm>true</vm>
+   * <env>flex</env>
+   * }
+   */
+  public abstract boolean isFlexCompat(@NotNull Project project, @NotNull DeploymentSource source);
+
+  /**
+   * @see AppEngineProjectService#isFlexCompat(Project, DeploymentSource).
+   */
+  public abstract boolean isFlexCompat(@NotNull Project project, @Nullable Artifact artifact);
+
+  /**
+   * Determines the {@link AppEngineEnvironment} type of the artifact.
+   */
+  @Nullable
+  public abstract AppEngineEnvironment getAppEngineArtifactEnvironment(@NotNull Project project,
+      @NotNull Artifact artifact);
+
+  /**
+   * Determines if a module has App Engine standard deployable artifacts. An artifact is considered
+   * deployable to App Engine standard if it has the AE facet and it is an exploded war artifact
+   * type.
+   */
+  public abstract boolean containsAppEngineStandardArtifacts(@NotNull Project project,
+      @NotNull Module module);
+
+  /**
+   * @see AppEngineProjectService#containsAppEngineStandardArtifacts(Project, Module).
+   */
+  public abstract boolean containsAppEngineStandardArtifacts(Project project,
+      Collection<Artifact> artifacts);
+
+  /**
+   * @see AppEngineProjectService#containsAppEngineStandardArtifacts(Project, Module).
+   */
+  public abstract boolean isAppEngineStandardArtifact(@NotNull Project project,
+      @NotNull Artifact artifact);
+
+  /**
+   * True if the artifact has the App Engine standard facet.
```

Comment:
Indirectly - artifacts are related to modules (in that they are produced by them); and modules have facets. I can update the comment to make it a bit clearer.